File: TasksWidget/widget/src/main/java/com/mitrejcevski/widget/dialog/NewGroupDialog.java
package com.mitrejcevski.widget.dialog;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.mitrejcevski.widget.R;
import com.mitrejcevski.widget.activity.GroupsListActivity;
import com.mitrejcevski.widget.database.DBManipulator;
import com.mitrejcevski.widget.model.Group;
import com.mitrejcevski.widget.provider.ListWidget;

public class NewGroupDialog extends DialogFragment {

    private static final String GROUP_ID_EXTRA = "group_id_extra";
    private Group mGroup;
    private View mView;

    public static NewGroupDialog newInstance(final int groupId) {
        NewGroupDialog dialog = new NewGroupDialog();
        Bundle args = new Bundle();
        args.putInt(GROUP_ID_EXTRA, groupId);
        dialog.setArguments(args);
        return dialog;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setCancelable(false);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final int id = getArguments().getInt(GROUP_ID_EXTRA);
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(R.string.groupEditorTitle);
        builder.setView(obtainView(id));
        addPositiveButton(builder);
        addNegativeButton(builder);
        return builder.create();
    }

    @SuppressLint("InflateParams")
    private View obtainView(final int groupId) {
        mView = getActivity().getLayoutInflater().inflate(R.layout.group_dialog_layout, null);
        if (groupId != -1) {
            mGroup = DBManipulator.INSTANCE.getGroupById(getActivity(), groupId);
            ((EditText) mView.findViewById(R.id.group_title_edit_text)).setText(mGroup.getGroupTitle());
        }
        return mView;
    }

    private void addPositiveButton(final AlertDialog.Builder builder) {
        builder.setPositiveButton(R.string.actionSave, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int i) {
                updateGroup();
                dialog.dismiss();
            }
        });
    }

    private void updateGroup() {
        EditText titleEdit = (EditText) mView.findViewById(R.id.group_title_edit_text);
        final String title = titleEdit.getText().toString();
        if (title.length() < 1)
            Toast.makeText(getActivity(), R.string.labelInvalidGroup, Toast.LENGTH_SHORT).show();
        else {
            completeSaving(title);
            ((GroupsListActivity) getActivity()).update();
        }
    }

    private void completeSaving(final String title) {
        if (mGroup == null)
            mGroup = new Group();
        mGroup.setGroupTitle(title);
        DBManipulator.INSTANCE.saveGroup(getActivity(), mGroup);
        Toast.makeText(getActivity(), R.string.successMessageLabel, Toast.LENGTH_SHORT).show();
        notifyWidget();
    }

    private void addNegativeButton(final AlertDialog.Builder builder) {
        builder.setNegativeButton(R.string.actionCancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int i) {
                dialog.dismiss();
            }
        });
    }

    private void notifyWidget() {
        final Intent fillInIntent = new Intent(getActivity(), ListWidget.class);
        fillInIntent.setAction(ListWidget.UPDATE_ACTION);
        getActivity().sendBroadcast(fillInIntent);
    }
}

File: widget/src/main/java/com/widget/ui/groups/NewGroupDialog.kt
package com.widget.ui.groups

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.view.View
import android.widget.EditText
import com.widget.R

internal class NewGroupDialog : DialogFragment() {

    private var callback: OnDoneCallback = EmptyCallback()
    private var layout: View? = null

    interface OnDoneCallback {
        fun onGroupReady(title: String)

        fun onGroupError()
    }

    fun withDoneCallback(callback: OnDoneCallback): NewGroupDialog {
        this.callback = callback
        return this
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isCancelable = false
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        layout = activity.layoutInflater.inflate(R.layout.group_dialog_layout, null)
        val builder = AlertDialog.Builder(activity)
        builder.setTitle(R.string.groupEditorTitle)
        builder.setView(layout)
        addPositiveButton(builder)
        addNegativeButton(builder)
        return builder.create()
    }

    private fun addPositiveButton(builder: AlertDialog.Builder) {
        builder.setPositiveButton(R.string.actionSave) { dialog, button ->
            updateGroup()
            dialog.dismiss()
        }
    }

    private fun updateGroup() {
        val titleEdit = layout?.findViewById(R.id.groupTitleEditText) as EditText
        val title = titleEdit.text.toString()
        if (title.length < 1) callback.onGroupError() else callback.onGroupReady(title)
    }

    private fun addNegativeButton(builder: AlertDialog.Builder) {
        builder.setNegativeButton(R.string.actionCancel) { dialog, button -> dialog.dismiss() }
    }

    private inner class EmptyCallback : OnDoneCallback {
        override fun onGroupReady(title: String) {

        }

        override fun onGroupError() {

        }
    }
}